fog/sample-paykit/src/client.rs

```diff
@@ -444,26 +444,35 @@ impl Client {
     /// self.ring_size elements.
     fn get_rings<T: RngCore + CryptoRng>(
         &mut self,
-        num_rings: usize,
+        true_inputs: &[(OwnedTxOut, TxOutMembershipProof)],
         rng: &mut T,
     ) -> Result<Vec<Vec<(TxOut, TxOutMembershipProof)>>> {
         mc_common::trace_time!(self.logger, "MobileCoinClient.get_rings");
 
+        let true_input_indices: HashSet<u64> = true_inputs
+            .iter()
+            .map(|input| input.0.global_index)
+            .collect();
+
+        let num_rings = true_inputs.len();
         let num_requested = num_rings * self.ring_size;
         let sample_limit = self.tx_data.get_global_txo_count() as usize;
 
-        // Randomly sample `num_requested` TxOuts, without replacement.
-        if sample_limit < num_requested {
+        // Randomly sample `num_requested` TxOuts, without replacement, not using
+        // true_inputs
+        if sample_limit < num_requested + true_inputs.len() {
             return Err(Error::InsufficientTxOutsInBlockchain(
                 sample_limit,
-                num_requested,
+                num_requested + true_inputs.len(),
             ));
         }
 
         let mut sampled_indices: HashSet<u64> = HashSet::default();
         while sampled_indices.len() < num_requested {
-            let index = rng.gen_range(0..sample_limit);
-            sampled_indices.insert(index as u64);
+            let index = rng.gen_range(0..sample_limit) as u64;
+            if !true_input_indices.contains(&index) {
```

Comment:
so, the reason I changed this code this way was, when running the test with a relatively small bootstrap, (which makes the test faster), I was getting problems where, my transactions were failing because the real input was being chosen as one of the real input's mixins. So the most obvious fix was to exclude all the real inputs from the sample. If we don't do something to prevent that, then the client can build an invalid transaction and the test can fail.

I think you may be right that a more subtle fix is better, where when sampling a ring for `x` we only exclude `x` as a mixin.

Since this is just used by the test client, I feel like it's not security critical. But to the extent that it's a model for real SDKs to crib from, I think your proposed change would be an improvement.

Reply:
Cool, I think we're on the same page.

Reply:
@samdealy I think the upshot here is that this code should get modified to test for real inputs. I'm fine if that happens in a follow-up PR, but please create an issue for this.